client/my-sites/marketplace/controller.js

```diff
@@ -8,25 +8,34 @@ import page from 'page';
  * Internal dependencies
  */
 import { getSiteFragment } from 'calypso/lib/route';
-import MarketplacePluginDetails from 'calypso/my-sites/marketplace/marketplace-plugin-details';
-import MarketplaceDomainUpsell from 'calypso/my-sites/marketplace/marketplace-domain-upsell';
-import MarketplacePluginSetup from 'calypso/my-sites/marketplace/marketplace-plugin-setup-status';
-import MarketplaceStandaloneThankYou from 'calypso/my-sites/marketplace/marketplace-stand-alone-thank-you';
-import MarketplaceTest from 'calypso/my-sites/marketplace/marketplace-test';
+import MarketplacePluginDetails from 'calypso/my-sites/marketplace/pages/marketplace-product-details';
+import MarketplaceDomainUpsell from 'calypso/my-sites/marketplace/pages/marketplace-domain-upsell';
+import MarketplacePluginSetup from 'calypso/my-sites/marketplace/pages/marketplace-plugin-setup-status';
+import MarketplaceStandaloneThankYou from 'calypso/my-sites/marketplace/pages/marketplace-stand-alone-thank-you';
+import MarketplaceTest from 'calypso/my-sites/marketplace/pages/marketplace-test';
+import { getDefaultProductInProductGroup } from 'calypso/my-sites/marketplace/marketplace-product-definitions';
 
-export function renderMarketplacePlugin( context, next ) {
-	const siteUrl = getSiteFragment( context.path );
-
-	const pluginSlug = decodeURIComponent( context.params.plugin );
+export function renderMarketplaceProduct( context, next ) {
+	const siteFragment = getSiteFragment( context.path );
+	const { productSlug: productSlugParam, productGroupSlug: productGroupSlugParam } = context.params;
+	let productSlug = productSlugParam ? decodeURIComponent( productSlugParam ) : null;
+	const productGroupSlug = productGroupSlugParam
+		? decodeURIComponent( productGroupSlugParam )
+		: null;
 
+	if ( ! siteFragment ) {
+		return page.redirect( '/home' );
```

Comment:
Wouldn't it be a better experience if we provide the site selection middleware here, and once the user selects a site, they're presented with the marketplace product detail page (given that the rest of the arguments are present)? This will present the user with the site selection middleware, but will redirect them to My Home, which to me sounds like a confusing experience. 

Also, since we already have the site selection middleware, is there any chance that this will ever run without a `siteFragment`?

Reply:
Yes I guess you are right, the site selection function looks to be doing a lot of things, I was not sure that was its main job. Will remove this condition! :)